mypyc/genops.py

```diff
@@ -494,46 +478,93 @@ def f(self, x: object) -> int: ...
+
+        func_info = FuncInfo()
+        self.func_infos.append(func_info)
+
         # If there is more than one environment in the environment stack, then we are visiting a
         # non-global function.
-        is_nested = len(self.environments) > 1
+        func_info.is_nested = len(self.environments) > 1
+        # If any of the blocks is another FuncDef, then this contains a nested function.
+        func_info.contains_nested = self.contains_func_def(fdef)
+        # func_info.contains_nested = any(isinstance(block, FuncDef) for block in fdef.body.body)
+
+        namespace = self.gen_func_ns()
 
         self.enter(fdef.name())
 
-        if is_nested:
-            # If this is a nested function, then add a 'self' field to the environment, since we
-            # will be instantiating the function as a method of a new class representing that
-            # original function.
-            self.environment.add_local(Var('self'), object_rprimitive, is_arg=True)
-        for arg in fdef.arguments:
-            assert arg.variable.type, "Function argument missing type"
-            self.environment.add_local(arg.variable, self.type_to_rtype(arg.variable.type),
-                                       is_arg=True)
+        if func_info.is_nested:
+            # Generate a callable class representing the function. The generated class is not yet
+            # functional, but used as a placeholder so that 'self' can use its RInstance type.
+            self.gen_callable_class(fdef, namespace)
+
+            # Add a 'self' field to the environment, since nested functions become the '__call__'
+            # class methods of callable classes. Use the callable class as the RInstance type.
+            self_target = self.environment.add_local_reg(Var('self'),
+                                                         RInstance(func_info.callable_class),
+                                                         is_arg=True)
+            func_info.self_reg = self.read_from_target(self_target, fdef.line)
+
+        # Regardless of whether the function contains a nested function, add arguments to local
+        # registers. These will be moved into an environment class if the function contains a
+        # nested function.
+        self.add_args_to_environment(fdef.arguments, fdef.line, local=True)
+
+        if func_info.contains_nested:
+            # Generate an instantiate an environment class now so that the compiler can refer to it
+            # while visiting nested functions. The attributes of the environment class will be
+            # populated while traversing the nested functions.
+            self.gen_env_class(fdef, namespace)
+            self.instantiate_env_class(fdef.line)
+
+            # Reiterate through the function arguments and replace local definitions (using
+            # registers) with references to the function's environment class.
+            self.add_args_to_environment(fdef.arguments, fdef.line, local=False)
```

Comment:
I think that the blocks of logic under `is_nested` and `contains_nested` should be pulled out into their own functions.

Reply:
Done.